### ISSUE TITLE
Please provide official binaries

### ISSUE THREAD

**Issue by ytrezq** [ledgerwatch/erigon#4153]
Many times building fails because of things like environment/compiler differences, especially in the case of Windows (Win64) or osx.

Providing build scripts is good but because of environment différences, they tend to fails.

**Comment by AskAlexSharov** [comment]
Support docker is easier: https://hub.docker.com/r/thorax/erigon

**Comment by ytrezq** [comment]
@AskAlexSharov for Linux, likely. That s why I m talking about Windows and osx.
And in my case, I need to pipe to Windows programs.

**Comment by AskAlexSharov** [comment]
1. In my head "Docker" - works on any host OS. On osx it works well. I believe on WSL2 docker works as well. 
2. logs - it's just text. if can't deal with stdin/stdout on win - you can write logs to file, then pipe file content where need. 
3. On win don't use "make" use "wmake.ps1"
4. if you provide binaries, why don't provide deb/brew/nix packages, etc... geth has some toolset for building: https://github.com/ethereum/go-ethereum/tree/master/build - but we never really supported it
5. maybe somebody will get ownership on it in future

**Comment by ytrezq** [comment]
@AskAlexSharov
> 1. In my head "Docker" - works on any host OS. On osx it works well. I believe on WSL2 docker works as well.

Right, the issue is if you need to interface the node with Windows or osx specific programs which is my case. On Linux there are several distributions and incompatibles Libraries. On Windows, the os comes a complete stack of backward compatible Microsoft s dll
> 2. logs - it's just text. if can't deal with stdin/stdout on win - you can write logs to file, then pipe file content where need.

I was talking about Windows pipes.
> 3. On win don't use "make" use "wmake.ps1"

It works, provided it match an unspecified installed toolchain version. Else lot of errors.
> 4. if you provide binaries, why don't provide deb/brew/nix packages, etc... geth has some toolset for building: https://github.com/ethereum/go-ethereum/tree/master/build - but we never really supported it

I do note you already have automated build scripts for automated pull request testing which could be used on purpose.
> 5. maybe somebody will get ownership on it in future

**Comment by AskAlexSharov** [comment]
“ backward compatible Microsoft s dll ” - does it include mingw runtime dll (Golang can’t link to msvc runtime)? libmdbx dll? Probably all our C dependencies (dozen) will create individual dll. It’s impossible to create statically-linked binary on  windows. Who will ensure that dll versions are correct on user’s machine? 

“already have automated build scripts for automated pull request testing” - and you just can’t imagine how much time, pain and blood took us to setup it for Windows. Weeks and 2 developers. 

I’m not complaining - just as I understand it’s big and likely unsolvable (no static-linking on win) task.

thank you for opening issue. 
If you like to do PR - you are welcome.

**Comment by ytrezq** [comment]
> “ backward compatible Microsoft s dll ” - does it include mingw runtime dll (Golang can’t link to msvc runtime)? libmdbx dll? Probably all our C dependencies (dozen) will create individual dll. It’s impossible to create statically-linked binary on windows. Who will ensure that dll versions are correct on user’s machine?

Disagree. With mingw, static linked is possible if the static versions of the Libraries are present. But it s about runtimes. Not things such as glibc versus musl. Those missing libraries are still common to all Windows versions.

> “already have automated build scripts for automated pull request testing” - and you just can’t imagine how much time, pain and blood took us to setup it for Windows. Weeks and 2 developers.

Correct, and since the work is almost done, it s not up the end users to spend almost so much time in order to deal  with this.
> I’m not complaining - just as I understand it’s big and likely unsolvable (no static-linking on win) task.
> 
> thank you for opening issue. If you like to do PR - you are welcome.

**Comment by AskAlexSharov** [comment]
If anybody willing to work on it - some link which maybe working: https://github.com/shadowmint/go-static-linking

**Comment by ytrezq** [comment]
@AskAlexSharov what about simply making erigon compatible with gcc-go? It would also solve this problem while bringing support to other architectures like mips and thus loongsun (a way explored by China for replacing x86).

**Comment by AskAlexSharov** [comment]
Maybe it’s good idea - I don’t have any opinion here.

**Comment by ytrezq** [comment]
@AskAlexSharov gcc-fo is only compatible with go up to 1.16. Is there lot of Erigon code concerned by this?

**Comment by AskAlexSharov** [comment]
- not much 
- some our deps (like anacrolix/torrent) already switched to go1.18 
- don't know if pprof and other go tools will work with gccgo or not

**Comment by ytrezq** [comment]
> * some our deps (like anacrolix/torrent) already switched to go1.18

Care must be made in order to ensure Erigon isn t using the versions affected. @AskAlexSharov does Erigon already rely on those versions?
> * don't know if pprof and other go tools will work with gccgo or not

Instead of using those specific tools you use the gcc and binutils tools used by all the other languages.

**Comment by AskAlexSharov** [comment]
https://goreleaser.com has https://github.com/goreleaser/goreleaser-cross

**Comment by ytrezq** [comment]
@AskAlexSharov gcc remains king for supported architectures (more than llvm). You won t get Erigon running in China and Russia with this.

**Comment by AskAlexSharov** [comment]
scope of this issue getting bigger and bigger: need research about cross-compilation and research gccgo and sign binaries and support all platforms in the world and configure CI for this and ...


cons of gccgo:
- not much people using it, no community (i'm about gccgo, not about gcc), way slower compilation, unclear support of goroutines 
- loose all tooling, or need re-check what will work: IntellijIdea? etc... "binutils tools used by all the other languages" - other languages don't have 100K goroutines - even if all that tools work need re-check them, update docs, etc...
- we have SilkWarm project - it's C++, it's already buildable by pure GCC and Clang. I think it's better solution than gccgo: https://erigon.substack.com/p/current-status-of-silkworm-and-silkrpc?s=r 

what done now:
- Docker image: https://hub.docker.com/r/thorax/erigon
- Makefile (wmake.ps1 provided for windows) does check pre-installed version of cc and go. then user can build us by `make erigon` 


I personally will try to distance from this issue: it's big, it's not related to core-functionality - size and usability of Archive node. Also long list of higher-priority things: SnapSync, Erigon2 upgrades, TheMerge, Hive integration, etc... I will drop link to this issue in our Discord chat - if somebody willing to work on it i'm fine (postpone it for several months also fine). 

Thank you for advises.

**Comment by ytrezq** [comment]
> cons of gccgo:
> not much people using it, no community (i'm about gccgo, not about gcc), way slower compilation, unclear support of goroutines.

Correct, it’s trading development time at the expense of slower user runtime (choosing faster to build instead of faster to run binaries).

> we have SilkWarm project - it's C++, it's already buildable by pure GCC and Clang. I think it's better solution than gccgo: https://erigon.substack.com/p/current-status-of-silkworm-and-silkrpc?s=r

Sounds like the way to go. Except of course it doesn’t runs nor not even support tracing of…

> Docker image: https://hub.docker.com/r/thorax/erigon

***Which don’t run Windows only programs inside it.***

> Makefile (wmake.ps1 provided for windows) does check pre-installed version of cc and go. then user can build us by make erigon

*** Just doesn’t works on a Fresh Windows 10 install with many errors like this one to be fixed :

PS C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06> .\wmake.ps1

    Avertissement de sécurité
    N’exécutez que des scripts que vous approuvez. Bien que les scripts en provenance d’Internet puissent être utiles, ce script est susceptible d’endommager votre ordinateur. Si
     vous approuvez ce script, utilisez l’applet de commande Unblock-File pour autoriser le script à s’exécuter sans ce message d’avertissement. Voulez-vous exécuter
    C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06\wmake.ps1 ?
    [N] Ne pas exécuter  [O] Exécuter une fois  [S] Suspendre  [?] Aide (la valeur par défaut est « N ») : O

     ------------------------------------------------------------------------------
      Erigon's wmake.ps1 : Selected target(s) erigon rpcdaemon sentry downloader integration
     ------------------------------------------------------------------------------

     Found Git version 2.36.1
     Erigon Branch :
     Erigon Tag    :
     Erigon Commit :

     Building erigon
    go: downloading github.com/ledgerwatch/log/v3 v3.4.1
    go: downloading github.com/ledgerwatch/erigon-lib v0.0.0-20220620073929-46bebb3317d9
    go: downloading github.com/urfave/cli v1.22.9
    go: downloading github.com/c2h5oh/datasize v0.0.0-20200825124411-48ed595a09d2
    go: downloading github.com/spf13/pflag v1.0.5
    go: downloading github.com/holiman/uint256 v1.2.0
    go: downloading github.com/anacrolix/go-libutp v1.2.0
    go: downloading github.com/anacrolix/log v0.13.1
    go: downloading github.com/anacrolix/torrent v1.44.0
    go: downloading golang.org/x/time v0.0.0-20220411224347-583f2d630306
    go: downloading github.com/spf13/cobra v1.4.0
    go: downloading golang.org/x/sys v0.0.0-20220503163025-988cb79eb6c6
    go: downloading google.golang.org/grpc v1.46.2
    go: downloading google.golang.org/protobuf v1.28.0
    go: downloading github.com/gofrs/flock v0.8.1
    go: downloading github.com/rs/cors v1.8.2
    go: downloading github.com/huin/goupnp v1.0.3
    go: downloading github.com/jackpal/go-nat-pmp v1.0.2
    go: downloading github.com/pion/stun v0.3.5
    go: downloading github.com/deckarep/golang-set v0.0.0-20180603214616-504e848d77ea
    go: downloading github.com/edsrzf/mmap-go v1.1.0
    go: downloading github.com/goccy/go-json v0.9.7
    go: downloading github.com/hashicorp/golang-lru v0.5.5-0.20210104140557-80c98217689d
    go: downloading golang.org/x/crypto v0.0.0-20220507011949-2cf3adece122
    go: downloading github.com/torquem-ch/mdbx-go v0.24.3-0.20220614090901-342411560dde
    go: downloading golang.org/x/exp v0.0.0-20220428152302-39d4317da171
    go: downloading github.com/go-stack/stack v1.8.1
    go: downloading github.com/mattn/go-colorable v0.1.11
    go: downloading github.com/mattn/go-isatty v0.0.14
    go: downloading github.com/VictoriaMetrics/metrics v1.18.1
    go: downloading golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
    go: downloading go.uber.org/atomic v1.9.0
    go: downloading github.com/cpuguy83/go-md2man/v2 v2.0.1
    go: downloading github.com/anacrolix/missinggo v1.3.0
    go: downloading github.com/anacrolix/mmsg v1.0.0
    go: downloading github.com/anacrolix/sync v0.4.0
    go: downloading github.com/ledgerwatch/secp256k1 v1.0.0
    go: downloading github.com/ugorji/go/codec v1.1.13
    go: downloading github.com/valyala/fastjson v1.6.3
    go: downloading github.com/grpc-ecosystem/go-grpc-middleware v1.3.0
    go: downloading github.com/google/btree v1.0.1
    go: downloading github.com/flanglet/kanzi-go v1.9.1-0.20211212184056-72dda96261ee
    go: downloading github.com/davecgh/go-spew v1.1.1
    go: downloading github.com/gorilla/websocket v1.5.0
    go: downloading github.com/inconshreveable/mousetrap v1.0.0
    go: downloading github.com/golang-jwt/jwt/v4 v4.4.1
    go: downloading github.com/json-iterator/go v1.1.12
    go: downloading github.com/xsleonard/go-merkle v1.1.0
    go: downloading github.com/pelletier/go-toml/v2 v2.0.2
    go: downloading github.com/RoaringBitmap/roaring v1.2.1
    go: downloading github.com/anacrolix/chansync v0.3.0
    go: downloading github.com/anacrolix/dht/v2 v2.16.2-0.20220311024416-dd658f18fd51
    go: downloading github.com/anacrolix/generics v0.0.0-20220510042907-b50562b436ec
    go: downloading github.com/anacrolix/missinggo/perf v1.0.0
    go: downloading github.com/anacrolix/missinggo/v2 v2.7.0
    go: downloading github.com/anacrolix/multiless v0.2.1-0.20211218050420-533661eef5dc
    go: downloading github.com/anacrolix/upnp v0.1.3-0.20220123035249-922794e51c96
    go: downloading github.com/bahlo/generic-list-go v0.2.0
    go: downloading github.com/dustin/go-humanize v1.0.0
    go: downloading github.com/lispad/go-generics-tools v1.0.0
    go: downloading github.com/pion/datachannel v1.5.2
    go: downloading github.com/pion/webrtc/v3 v3.1.24-0.20220208053747-94262c1b2b38
    go: downloading github.com/pkg/errors v0.9.1
    go: downloading github.com/pelletier/go-toml v1.9.5
    go: downloading github.com/golang/snappy v0.0.4
    go: downloading github.com/spaolacci/murmur3 v1.1.0
    go: downloading github.com/emirpasic/gods v1.18.1
    go: downloading github.com/stretchr/testify v1.7.2
    go: downloading github.com/VictoriaMetrics/fastcache v1.10.0
    go: downloading github.com/emicklei/dot v0.16.0
    go: downloading github.com/russross/blackfriday/v2 v2.1.0
    go: downloading github.com/valyala/histogram v1.2.0
    go: downloading github.com/huandu/xstrings v1.3.2
    go: downloading github.com/bradfitz/iter v0.0.0-20191230175014-e8f45d346db8
    go: downloading github.com/anacrolix/envpprof v1.2.1
    go: downloading github.com/ajwerner/btree v0.0.0-20211221152037-f427b3e689c0
    go: downloading github.com/tidwall/btree v0.7.2-0.20211211132910-4215444137fc
    go: downloading github.com/pion/logging v0.2.2
    go: downloading github.com/pion/sctp v1.8.2
    go: downloading github.com/rs/dnscache v0.0.0-20210201191234-295bba877686
    go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
    go: downloading github.com/modern-go/reflect2 v1.0.2
    go: downloading github.com/tendermint/go-amino v0.14.1
    go: downloading github.com/tendermint/tendermint v0.31.11
    go: downloading github.com/valyala/fastrand v1.1.0
    go: downloading github.com/dop251/goja v0.0.0-20211011172007-d99e4b8cbf48
    go: downloading github.com/cespare/xxhash/v2 v2.1.2
    go: downloading github.com/benbjohnson/immutable v0.3.0
    go: downloading github.com/pmezard/go-difflib v1.0.0
    go: downloading gopkg.in/yaml.v3 v3.0.1
    go: downloading github.com/pion/dtls/v2 v2.1.2
    go: downloading github.com/pion/interceptor v0.1.7
    go: downloading github.com/pion/ice/v2 v2.1.20
    go: downloading github.com/pion/randutil v0.1.0
    go: downloading github.com/pion/rtcp v1.2.9
    go: downloading github.com/pion/rtp v1.7.4
    go: downloading github.com/pion/sdp/v3 v3.0.4
    go: downloading github.com/pion/srtp/v2 v2.0.5
    go: downloading github.com/pion/transport v0.13.0
    go: downloading golang.org/x/net v0.0.0-20220425223048-2871e0cb64e4
    go: downloading github.com/anacrolix/stm v0.3.0
    go: downloading google.golang.org/genproto v0.0.0-20200825200019-8632dd797987
    go: downloading github.com/golang/protobuf v1.5.2
    go: downloading github.com/dlclark/regexp2 v1.4.1-0.20201116162257-a2a8dda75c91
    go: downloading golang.org/x/text v0.3.7
    go: downloading github.com/go-sourcemap/sourcemap v2.1.3+incompatible
    go: downloading github.com/pion/turn/v2 v2.0.6
    go: downloading github.com/pion/mdns v0.0.5
    go: downloading github.com/google/uuid v1.3.0
    go: downloading github.com/btcsuite/btcd v0.22.0-beta
    go: downloading github.com/gogo/protobuf v1.3.2
    go: downloading github.com/go-logfmt/logfmt v0.5.0
    go: downloading github.com/go-kit/kit v0.10.0
    go: downloading github.com/pion/udp v0.1.1
    go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1
    cmd\downloader\trackers\embed.go:17:12: pattern trackerslist/trackers_all_http.txt: no matching files found
    turbo\snapshotsync\snapshothashes\embed.go:34:12: pattern erigon-snapshots/bor-mainnet.toml: no matching files found

     Built erigon. Run C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06\build\bin\erigon.exe to launch

     Building rpcdaemon
    turbo\snapshotsync\snapshothashes\embed.go:34:12: pattern erigon-snapshots/bor-mainnet.toml: no matching files found

     Built rpcdaemon. Run C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06\build\bin\rpcdaemon.exe to launch

     Building sentry
    cmd\downloader\trackers\embed.go:17:12: pattern trackerslist/trackers_all_http.txt: no matching files found
    turbo\snapshotsync\snapshothashes\embed.go:34:12: pattern erigon-snapshots/bor-mainnet.toml: no matching files found

     Built sentry. Run C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06\build\bin\sentry.exe to launch

     Building downloader
    cmd\downloader\trackers\embed.go:17:12: pattern trackerslist/trackers_all_http.txt: no matching files found
    turbo\snapshotsync\snapshothashes\embed.go:34:12: pattern erigon-snapshots/bor-mainnet.toml: no matching files found

     Built downloader. Run C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06\build\bin\downloader.exe to launch

     Building integration
    turbo\snapshotsync\snapshothashes\embed.go:34:12: pattern erigon-snapshots/bor-mainnet.toml: no matching files found

     Built integration. Run C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06\build\bin\integration.exe to launch
    PS C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06> C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06\build\bin\erigon.exe
    C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06\build\bin\erigon.exe : Le terme «
    C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8wekyb3d8bbwe\TempState\Downloads\erigon-2022.06.06\build\bin\erigon.exe» n'est pas reconnu comme nom d'applet de commande, fonction, fichier de
    script ou programme exécutable. Vérifiez l'orthographe du nom, ou si un chemin d'accès existe, vérifiez que le chemin d'accès est correct et réessayez.
    Au caractère Ligne:1 : 1
    + C:\Users\example\AppData\Local\Packages\Microsoft.MicrosoftEdge_8weky ...
    + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : ObjectNotFound: (C:\Users\exampl...\bin\erigon.exe:String) [], CommandNotFoundException
        + FullyQualifiedErrorId : CommandNotFoundException

@AskAlexSharov : it shouldn’t be the user’s issue instead of the developer’s issue to get Erigon available on `\\.\pipe\geth.ipc`

**Comment by AskAlexSharov** [comment]
“ trackers_all_http.txt: no matching files found” we will fix, it’s about update git submodules.

Ipc - erigon doesn’t support ipc.

**Comment by AskAlexSharov** [comment]
1. Fixed by https://github.com/ledgerwatch/erigon/pull/4532

**Comment by ytrezq** [comment]
> “ trackers_all_http.txt: no matching files found” we will fix, it’s about update git submodules.

That’s far from being the single problem. @AskAlexSharov it shouldn’t be the user’s issue instead of the developer’s issue to get the rpcdaemon available on `\\.\pipe\geth.ipc`. I don’t mind about a Windows based docker but it has to be provided.
> 
> Ipc - erigon doesn’t support ipc.

Sorry, I was meaning the rpcdaemon of course.

**Comment by AskAlexSharov** [comment]
Docker: https://hub.docker.com/r/thorax/erigon
If something doesn't work - please create an issue

**Comment by AskAlexSharov** [comment]
In perfect world: wouldn’t it be simpler to be always healthy and rich?

I'm totally agree with everything you say, but it's low-priority very-big task. High priority: archive node sync speed and size, the-merge support. 

"that version of docker doesn’t let use Windows pipes" - I don't understand the problem, create issue with command and error message we can't fix it.

**Comment by ytrezq** [comment]
> "that version of docker doesn’t let use Windows pipes" - I don't understand the problem, create issue with command and error message we can't fix it.

The image contains a Linux os. Not Windows.

**Comment by mandrigin** [comment]
> Instead of opening 200 to 300 issues mainly because of incompatible versionning build systems

In theory, yes, in practice supporting and updating all possible versions for different OSes and for different linux distros, flavours and versions is a huge job.

I will welcome if someone runs and supports an unofficial package for Windows or something. Unlikely we will do this anytime soon though. Not many from the core contributors use Windows.

Docker is the best way to run unless you want to contribute or debug something. So far we provide linux/amd64 that fits if you want to run it on most public and private cloud location and compatible with cluster management systems.

**Comment by ytrezq** [comment]
> > Instead of opening 200 to 300 issues mainly because of incompatible versionning build systems
> 
> In theory, yes, in practice supporting and updating all possible versions for different OSes and for different linux distros, flavours and versions is a huge job.
> 
> I will welcome if someone runs and supports an unofficial package for Windows or something. Unlikely we will do this anytime soon though. Not many from the core contributors use Windows.
> 
> Docker is the best way to run unless you want to contribute or debug something. So far we provide linux/amd64 that fits if you want to run it on most public and private cloud location and compatible with cluster management systems.

@mandrigin this doesn’t have value with Windows where unless you have to use Reactos in russia, the only distributor is Microsoft.

**Comment by ytrezq** [comment]
> Then I am lost in what your point is :).

@mandrigin my point is I need a Windows binary, if you didn t read the whole thread.
The logic behind, is I m failing to do it for a single version, so seek core team to update their build integration like I was accustomed to with Openethereum in order to not only have code but something which can be used.

**Comment by AskAlexSharov** [comment]
« when using gccgo toolchain» - never used gccgo. And it doesn’t support go1.18 

So, please don’t use gccgo

**Comment by ytrezq** [comment]
@AskAlexSharov that’s the point of the message above : I’m using the official go. Anyway, as you can the errors above are C++ compiler errors. Not golang ones.

**Comment by AskAlexSharov** [comment]
@ytrezq then bug-report must contain some related context: 
go env
cc --version
uname -a

from https://github.com/ledgerwatch/erigon#system-requirements - Golang version >= 1.18 , GCC 10+

**Comment by mandrigin** [comment]
> @AskAlexSharov that’s the point of the message above : I’m using the official go. Anyway, as you can the errors above are C++ compiler errors. Not golang ones.

If you are using Ubuntu, they don’t ship proper Go version, it is something like 1.15 there, when the official Go release is 1.19. And I see most of the errors you shared are from pre-generics Go version.

But yeah, as Alex says, provide your build config.

**Comment by ytrezq** [comment]
@AskAlexSharov
> from https://github.com/ledgerwatch/erigon#system-requirements - Golang version >= 1.18 , GCC 10+

    [liveuser@localhost-live erigon]$ go env
    cc --version
    uname -a
    GO111MODULE=""
    GOARCH="amd64"
    GOBIN=""
    GOCACHE="/home/liveuser/.cache/go-build"
    GOENV="/home/liveuser/.config/go/env"
    GOEXE=""
    GOEXPERIMENT="fieldtrack,regabiwrappers"
    GOFLAGS=""
    GOHOSTARCH="amd64"
    GOHOSTOS="linux"
    GOINSECURE=""
    GOMODCACHE="/home/liveuser/go/pkg/mod"
    GONOPROXY=""
    GONOSUMDB=""
    GOOS="linux"
    GOPATH="/home/liveuser/go"
    GOPRIVATE=""
    GOPROXY="https://proxy.golang.org,direct"
    GOROOT="/usr"
    GOSUMDB="sum.golang.org"
    GOTMPDIR=""
    GOTOOLDIR="/usr/libexec/gcc/x86_64-redhat-linux/12"
    GOVCS=""
    GOVERSION="unknown"
    GCCGO="/usr/bin/gccgo"
    GOAMD64="v1"
    AR="gcc-ar"
    CC="gcc"
    CXX="g++"
    CGO_ENABLED="1"
    GOMOD="/tmp/erigon/go.mod"
    GOWORK=""
    CGO_CFLAGS="-g -O2"
    CGO_CPPFLAGS=""
    CGO_CXXFLAGS="-g -O2"
    CGO_FFLAGS="-g -O2"
    CGO_LDFLAGS="-g -O2"
    PKG_CONFIG="pkg-config"
    GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1474644251=/tmp/go-build -gno-record-gcc-switches -funwind-tables"
    cc (GCC) 12.2.1 20220819 (Red Hat 12.2.1-1)
    Copyright © 2022 Free Software Foundation, Inc.
    Ce logiciel est un logiciel libre; voir les sources pour les conditions de copie.  Il n'y a
    AUCUNE GARANTIE, pas même pour la COMMERCIALISATION ni L'ADÉQUATION À UNE TÂCHE PARTICULIÈRE.
    
    Linux localhost-live 5.10.7-201.fc34.x86_64 #1 SMP Mon Jan 18 14:44:51 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
    [liveuser@localhost-live erigon]$

**Comment by mandrigin** [comment]
> > If you are using Ubuntu, they don’t ship proper Go version, it is something like 1.15 there, when the official Go release is 1.19. And I see most of the errors you shared are from pre-generics Go version.
> > But yeah, as Alex says, provide your build config.
> 
> @mandrigin : not an issue with a full up to date Fedora 37.

what do `go version` show?

this part is weird
```
GOVERSION="unknown"
```

**Comment by mandrigin** [comment]
```
error: use of undefined type 'T'
   89 | type FreeListG[T any] struct {
```

This error shows that your Golang doesn't support generics. Generics were introduced in go 1.18.

**Comment by ytrezq** [comment]
@mandrigin :

    [liveuser@localhost-live erigon]$ go version
    go version unknown linux/amd64

the package manager is listing version 1.19.

**Comment by mandrigin** [comment]
> @mandrigin :
> 
> ```
> [liveuser@localhost-live erigon]$ go version
> go version unknown linux/amd64
> ```
> 
> the package manager is listing version 1.19.
> 
> You don’t provide binaries in the name of multiple environment, but please realize the problem is even worst for compiling.

I’ve don’t provide official binaries for now because we are a small team and we had other things to focus on. We are not against that in principal, it just needs some attention and work that we don’t have.

“Version unknown” is suspicious, maybe the package manager is broken or something. I can easily compile on both go 1.18 and 1.19 on Mac and Linux alike.

**Comment by revitteth** [comment]
Please check latest release, which has binaries attached :) https://github.com/ledgerwatch/erigon/releases/tag/v2.39.0
